Implemented a Spring-based module, which handles gym CRM system

NazzP opened the pull request:
## Summary
Used spring-core, web, context for configuration. For tests were used JUnit5 & Mockito's libraries. Also tomcat and jackson dependencies for server and convertion implementation.

- This project represents a 3-tier application with facade pattern and storage implementation by itself.
- For testing request was used Postman application ( collection also attached in source root ) 
- For testing was implemented Unit & Integration Tests | besides, configuration and utils was tested also
-  Username going to be calculated from Trainer/Trainee first name and last name
by concatenation by using dot as a separator (eg. John.Smith)
- In the case that already exists Trainer or Trainee with the same pair of first and
last name as a suffix to the username should be added a serial number.
- Password should be generated as a random 10 chars length string

## Run application
- **mvn clean package** to build a .war file
- **docker build -t gymcrmsystem .** to build the image
- **docker run -p 8080:8080 gymcrmsystem** to run a container

## Stop application
- **docker ps** to display list of containers
- **docker stop <Container ID>** to stop a contrainer

## Checklist
- Coverage ( mvn clean verify with jacoco plugin ):

![зображення](https://github.com/user-attachments/assets/8f4cb905-2386-4d10-8461-c53ce5d26e19)
![зображення](https://github.com/user-attachments/assets/652db66f-61cf-42e2-a10e-8d7f40c583e6)

- mvn clean install -DskipTests ran successfully without issues:

![зображення](https://github.com/user-attachments/assets/afdcc263-fe9f-462d-aac4-e5a0cfc059c5)

- mvn clean install ran successfully without issues:

![зображення](https://github.com/user-attachments/assets/66dbccb0-6410-4a7e-be50-a2107357afed)

NazzP committed:
Created core entity models: Trainee, Trainer, Training, and TrainingType

- Implemented `User` abstract class as a base for user-related entities.
- Created `Trainee` and `Trainer` models extending `User`, with additional fields for `dateOfBirth`, `address`, and `specialization`.
- Added `Training` model with `Long` references for `traineeId`, `trainerId`, and related training details.
- Introduced `TrainingType` model to represent different types of training.

NazzP committed:
Added DTO classes for User, Trainee, Trainer, Training, and TrainingType

 Implemented `UserDTO`, `TraineeDTO`, `TrainerDTO`, `TrainingDTO`, and `TrainingTypeDTO` for data transfer.
- Each DTO class provides `fromEntity` and `toEntity` methods for easy conversion between entity models and their DTO counterparts.
- All DTO classes are annotated with `@Value`, ensuring immutability and thread-safety.

NazzP committed:
Add configuration classes for managing entity storage (Training, Trainer, Trainee)

- Created `TrainingStorageConfiguration`
- Created `TrainerStorageConfiguration`
- Created `TraineeStorageConfiguration`

All of them provide storage beans and counters for Training, Trainer, and Trainee

NazzP committed:
Added DAO interfaces and implementations

Corresponding implementations for CRUD operations

NazzP committed:
Added custom exceptions: ObjectNotFoundException and NullObjectReferenceException

ObjectNotFoundException -  is thrown when an object that is expected to be found in a data but not located
NullObjectReferenceException - is thrown when an object is expected to be non-null but is found to be null

NazzP committed:
Implemented service layers for Training, Trainer, and Trainee

- Added `TrainingService`, `TrainerService`, and `TraineeService` interfaces with corresponding implementations.

NazzP committed:
Implement REST controllers for Trainee, Trainer, and Training entities

- Controllers include methods for creating, retrieving, and updating entities:
  - `TraineeController`: `/api/trainees` for creating, retrieving, updating, and deleting trainees.
  - `TrainerController`: `/api/trainers` for creating, retrieving, and updating trainers.
  - `TrainingController`: `/api/trainings` for creating and retrieving training sessions.

NazzP committed:
Create GymCrmSystemApplication.java

NazzP committed:
Create GlobalExceptionHandler.java

NazzP committed:
Implemented Facade pattern for controller services

NazzP committed:
Refactor Storages: Replaces configuration-based storage with component-based service

NazzP committed:
Refactor: Split Facade into Interface and Implementation

NazzP committed:
Refactor: Renamed DAO to Repository

NazzP committed:
Refactor: Moved Entity Conversion Logic to Mapper and Renamed DTO to Dto

NazzP committed:
Docs: Added Documentation to Services

NazzP committed:
Changes up to changed Dto to Mapper

NazzP committed:
Refactor: Rename Classes and Variables to Follow Naming Conventions

NazzP committed:
Refactor whole application : ( without using spring-boot )

NazzP committed:
Renamed exceptions

NazzP committed:
Service tests impl

NazzP committed:
Create UsernameGenerator.java

NazzP committed:
Update UsernameGenerator.java

NazzP committed:
added findAll() to storages

NazzP committed:
documentations upd

NazzP committed:
EntityAlreadyExistsException impl

NazzP committed:
Impl methods in repositories

- saveNew for creation new User
- existsByUsername for checking if user with specific username exists

NazzP committed:
Create UserRepository.java

NazzP committed:
Username ckecker and auto config implemented in service

NazzP committed:
docuimentation upd

small changes
<p> using on breaks

NazzP committed:
deleted saveNew method

- Now, catching exception on Service layer using findById

NazzP committed:
Using Generic strorage instead of specific for Entity

NazzP committed:
parser Impl

NazzP committed:
Logged the application

NazzP committed:
added Integration Tests

NazzP committed:
upd Appl Tests fro checking beans

Using List for checking all needed

NazzP committed:
UsernameGeneratorTest Impl

- also refactored DAO and Service Tests

NazzP committed:
Added Config tests

NazzP committed:
password generation Impl

NazzP committed:
Create GymCRMSystem.postman_collection.json

NazzP committed:
pom.xml

Anderwerty commented on pom.xml:
```diff
@@ -0,0 +1,151 @@
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>3.4.1</version>
+        <relativePath/> <!-- lookup parent from repository -->
```
please remove this tag and redundant comment

Anderwerty commented on pom.xml:
```diff
@@ -0,0 +1,151 @@
+    <version>0.0.1-SNAPSHOT</version>
+    <name>Gym-CRM-System</name>
+    <description>Gym-CRM-System</description>
+    <url/>
```
please remove redundant tags

Anderwerty commented on pom.xml:
```diff
@@ -0,0 +1,151 @@
+        <java.version>17</java.version>
+    </properties>
+    <dependencies>
+        <dependency>
```
sorry, why not use starts and dependencyManager from parent pom.xml

Anderwerty commented on pom.xml:
```diff
@@ -0,0 +1,151 @@
+    </plugins>
+</build>
+
+        </project>
```
formatting issue: this tag was shifted

Anderwerty commented on src/main/java/org/example/gymcrmsystem/config/AppConfig.java:
```diff
@@ -0,0 +1,11 @@
+@ComponentScan(basePackages = "org.example.gymcrmsystem")
+@PropertySource("classpath:application.properties")
+public class AppConfig {
+}
```
please add a blank line

Anderwerty commented on src/main/java/org/example/gymcrmsystem/config/JacksonConfig.java:
```diff
@@ -0,0 +1,40 @@
+@Configuration
+public class JacksonConfig {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JacksonConfig.class);
```
why do we need logger here?

Anderwerty commented on src/main/java/org/example/gymcrmsystem/controller/TrainerController.java:
```diff
@@ -0,0 +1,35 @@
+    private final TrainerFacade trainerFacade;
+
+    @Autowired
+    public TrainerController(TrainerFacade trainerFacade) {
```
why not use lombok for constructor generation?

Anderwerty commented on src/main/java/org/example/gymcrmsystem/controller/TrainerController.java:
```diff
@@ -0,0 +1,35 @@
+    public ResponseEntity<TrainerDto> updateTrainer(@PathVariable("id") Long id, @RequestBody TrainerDto trainerDTO) {
+        return new ResponseEntity<>(trainerFacade.updateTrainer(id, trainerDTO), HttpStatus.OK);
+    }
+}
```
please add a blank line

Anderwerty commented on src/main/java/org/example/gymcrmsystem/dto/TraineeDto.java:
```diff
@@ -0,0 +1,22 @@
+import java.io.Serializable;
+import java.util.Date;
+
+@Data
```
this annotation also provides us toStrign method. But we should not log personal information
https://www.baeldung.com/lombok-tostring

Anderwerty commented on src/main/java/org/example/gymcrmsystem/dto/TraineeDto.java:
```diff
@@ -0,0 +1,22 @@
+@AllArgsConstructor
+@Builder
+public class TraineeDto implements Serializable, Identifiable<Long> {
+    Long id;
```
what about the access modifier? why not to make all fields private?

Anderwerty commented on src/main/java/org/example/gymcrmsystem/dto/TrainerDto.java:
```diff
@@ -0,0 +1,21 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class TrainerDto implements Serializable, Identifiable<Long> {
```
the same here

Anderwerty commented on src/main/java/org/example/gymcrmsystem/mapper/TraineeMapper.java:
```diff
@@ -0,0 +1,35 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class TraineeMapper {
```
Let's use mapstruct for mapper creation
https://www.baeldung.com/mapstruct

Anderwerty commented on src/main/java/org/example/gymcrmsystem/parser/JsonStorageParser.java:
```diff
@@ -0,0 +1,55 @@
+@Component
+public class JsonStorageParser<K, T extends Identifiable<K>> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JsonStorageParser.class);
```
We can use Lombok to provide us with an instance of logger https://projectlombok.org/features/log

Anderwerty commented on src/main/java/org/example/gymcrmsystem/parser/JsonStorageParser.java:
```diff
@@ -0,0 +1,55 @@
+    private final ObjectMapper objectMapper;
+
+    public JsonStorageParser() {
+        this.objectMapper = new ObjectMapper();
```
why not use injection here?

NazzP committed:
resolved comments

- upd pom.xml ( using parent version + dependency managment )
- logger impl using annotation
- lombok using for constructor generation in Controllers
- toString excluding personal info
- access modifier for fields ( 'private' ) impl in Dto's
- mapstruct impl

NazzP committed:
resolved meeting comments

- using properties for versioning dependencies
- lombok scope 'provided'
- import without wildcard
- @RequiredArgsConstructor instead of writing constuctor
- logger instance to upper case
- value ingection in constructor
- storage @Component deleted ( cz we using @Bean for every storage )
- passwordGenerator - using bean of this class instead of static
- ToString on fields
- jacoco pluugin impl and tested coverage

NazzP committed:
controllers deleted

NazzP committed:
DockerFile to run project on Tomcat server
